artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java

```diff
@@ -467,6 +500,28 @@ public static void validateProtocolData(ReferenceIDSupplier referenceIDSupplier,
       }
    }
 
+   /**
+    * Checks if the message ref should be filtered or not.
+    * @param ref the message to check
+    * @return true if the INTERNAL_RECEIVER_ID_FILTER annotation of the message is set to a different value
+    * than the remoteMirrorID, false otherwise.
+    */
+   public boolean filterMessage(MessageReference ref) {
```

Comment:
Whilst this may stop the message actually being sent, its not clear to me it does anything around actually cleaning it up? As far as the broker is concerned it sent a message, but the other end can never acknowledge it since it never actually arrives as it wasnt sent...so is it just stuck in the structure as a delivered-but-not-acked message? Presumably the snf queue delivering metrics would show this (and and be forever incorrect) if so.

Reply:
I added another commit trying another approach at filtering by doing it at an higher level. Let me know what you think about this.